site/content/blog/carvelizing-helm-chart.md

```diff
@@ -0,0 +1,296 @@
+$ wget https://raw.githubusercontent.com/vmware-tanzu/carvel-kapp-controller/83fffcfe99a65031b4170813acf94f8d5058b346/hack/install-deps.sh
+$ chmod a+x ./install-deps.sh
+$ ./install-deps.sh
+$ kubectl apply -f https://github.com/vmware-tanzu/carvel-kapp-controller/releases/latest/download/release.yml
+```
+
+**Note**: Going forward, I will be using [`kapp`](https://carvel.dev/kapp/) to deploy the yaml. `kapp` is a cli used to deploy and view groups of Kubernetes resources as “application”.
+
+--------------------------
+
+## Authoring a Carvel Package
+
+To create a package, we need to create two CRs. We will go through step by step to author an nginx helm chart:
+
+**1. Create Package Metadata**: Package Metadata contains very high level information and description about the package. Multiple versions of a package share same package metadata.
+
+```bash
+
+$ cat <<EOF > pkgMetadata.yaml
+apiVersion: data.packaging.carvel.dev/v1alpha1
+kind: PackageMetadata
+metadata:
+  # This will be the name of our package metadata
+  name: nginx.bitnami.vmware.com
+spec:
+  displayName: "Bitnami Nginx Carvel Package"
+  longDescription: "Proxifying Server"
+  shortDescription: "Proxifying Server"
+  categories:
+  - proxy-server
+  providerName: VMWare
+  maintainers:
+  - name: "Carvel"
+  - name: "CarvelInd"
+  - name: "Rohit Aggarwal"
+EOF
+```
+
+**2. Package Helm Chart**: Package is a combination of configuration metadata and OCI images that informs the package manager what software it holds and how to install itself onto a Kubernetes cluster.
+
+```bash
+
+$ cat <<EOF > 1.0.0.yaml
+apiVersion: data.packaging.carvel.dev/v1alpha1
+kind: Package
+metadata:
+  # Must be of the form '<spec.refName>.<spec.version>' (Note the period)
+  name: nginx.bitnami.vmware.com.1.0.0
+spec:
+  # The name of the PackageMetadata associated with this version
+  # Must be a valid PackageMetadata name (see PackageMetadata CR for details)
+  # Cannot be empty
+  refName: nginx.bitnami.vmware.com
+  # Package version; Referenced by PackageInstall;
+  # Must be valid semver (required)
+  # Cannot be empty
+  version: 1.0.0
+  # Version release notes (optional; string)
+  releaseNotes:
+    Initial release of the nginx package by wrapping helm Chart. Nginx Helm chart version is 9.5.4
+  # valuesSchema can be used to show template values that
+  # can be configured by users when a Package is installed.
+  # These values should be specified in an OpenAPI schema format. (optional)
+  # For helm chart, we can either define the configurable values here or let it be. Even if we dont define them here, we can still customize them.
+  valuesSchema:
+    openAPIv3:
+      title: nginx.bitnami.vmware.com
+      examples:
+      properties:
+  template:
+    spec:
+      fetch:
+      - helmChart:
+          name: nginx
+          version: 9.5.4
+          repository:
+            # From where to pull the helm chart
+            url: https://charts.bitnami.com/bitnami 
```

Comment:
we should call out that this is one way of doing this. another way is to use imgpkg bundle to store helm chart itself.

Reply:
I think I follow what DK is saying above but honestly I'm pretty dumb and it would be helpful to see both alternatives as fully formed examples,  i.e. show me both options  rather than just "call out" the other option.

Reply:
Adopted. Added a Note in the end. `This is one way of packaging helm chart. Another way is to use `imgpkg` bundle to store the helm chart itself and then reference it. `